File: springboot-export-pdf/src/main/java/com/knf/dev/util/PDFGenerator.java
package com.knf.dev.util;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.util.List;
import java.util.stream.Stream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.itextpdf.text.BaseColor;
import com.itextpdf.text.Chunk;
import com.itextpdf.text.Document;
import com.itextpdf.text.DocumentException;
import com.itextpdf.text.Element;
import com.itextpdf.text.Font;
import com.itextpdf.text.FontFactory;
import com.itextpdf.text.Paragraph;
import com.itextpdf.text.Phrase;
import com.itextpdf.text.pdf.PdfPCell;
import com.itextpdf.text.pdf.PdfPTable;
import com.itextpdf.text.pdf.PdfWriter;
import com.knf.dev.model.Employee;

public class PDFGenerator {

	private static Logger logger = LoggerFactory.getLogger(PDFGenerator.class);

	public static ByteArrayInputStream employeePDFReport(List<Employee> employees) {
		Document document = new Document();
		ByteArrayOutputStream out = new ByteArrayOutputStream();

		try {

			PdfWriter.getInstance(document, out);
			document.open();

			// Add Text to PDF file ->
			Font font = FontFactory.getFont(FontFactory.COURIER, 14, BaseColor.BLACK);
			Paragraph para = new Paragraph("Employee Table", font);
			para.setAlignment(Element.ALIGN_CENTER);
			document.add(para);
			document.add(Chunk.NEWLINE);

			PdfPTable table = new PdfPTable(3);
			// Add PDF Table Header ->
			Stream.of("ID", "First Name", "Last Name").forEach(headerTitle -> {
				PdfPCell header = new PdfPCell();
				Font headFont = FontFactory.getFont(FontFactory.HELVETICA_BOLD);
				header.setBackgroundColor(BaseColor.LIGHT_GRAY);
				header.setHorizontalAlignment(Element.ALIGN_CENTER);
				header.setBorderWidth(2);
				header.setPhrase(new Phrase(headerTitle, headFont));
				table.addCell(header);
			});

			for (Employee employee : employees) {
				PdfPCell idCell = new PdfPCell(new Phrase(employee.getId().toString()));
				idCell.setPaddingLeft(4);
				idCell.setVerticalAlignment(Element.ALIGN_MIDDLE);
				idCell.setHorizontalAlignment(Element.ALIGN_CENTER);
				table.addCell(idCell);

				PdfPCell firstNameCell = new PdfPCell(new Phrase(employee.getFirstName()));
				firstNameCell.setPaddingLeft(4);
				firstNameCell.setVerticalAlignment(Element.ALIGN_MIDDLE);
				firstNameCell.setHorizontalAlignment(Element.ALIGN_LEFT);
				table.addCell(firstNameCell);

				PdfPCell lastNameCell = new PdfPCell(new Phrase(String.valueOf(employee.getLastName())));
				lastNameCell.setVerticalAlignment(Element.ALIGN_MIDDLE);
				lastNameCell.setHorizontalAlignment(Element.ALIGN_RIGHT);
				lastNameCell.setPaddingRight(4);
				table.addCell(lastNameCell);
			}
			document.add(table);

			document.close();
		} catch (DocumentException e) {
			logger.error(e.toString());
		}

		return new ByteArrayInputStream(out.toByteArray());
	}
}

File: kotlin-spring-main/KotlinSpringExportPDF/src/main/kotlin/com/knf/dev/KotlinSpringExportPDF/util/PDFGenerator.kt
package com.knf.dev.KotlinSpringExportPDF.util

import com.itextpdf.text.*
import com.itextpdf.text.pdf.PdfPCell
import com.itextpdf.text.pdf.PdfPTable
import com.itextpdf.text.pdf.PdfWriter
import com.knf.dev.KotlinSpringExportPDF.model.Employee
import org.slf4j.LoggerFactory
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.util.stream.Stream


object PDFGenerator {
    private val logger = LoggerFactory.getLogger(PDFGenerator::class.java)
    fun employeePDFReport(employees: List<Employee>): ByteArrayInputStream {
        val document = Document()
        val out = ByteArrayOutputStream()
        try {
            PdfWriter.getInstance(document, out)
            document.open()

            // Add Text to PDF file ->
            val font = FontFactory.getFont(
                FontFactory.COURIER, 14f,
                BaseColor.BLACK
            )
            val para = Paragraph("Employee Table", font)
            para.alignment = Element.ALIGN_CENTER
            document.add(para)
            document.add(Chunk.NEWLINE)
            val table = PdfPTable(3)
            // Add PDF Table Header ->
            Stream.of("ID", "First Name", "Last Name").forEach { headerTitle: String? ->
                val header = PdfPCell()
                val headFont = FontFactory.getFont(FontFactory.HELVETICA_BOLD)
                header.backgroundColor = BaseColor.LIGHT_GRAY
                header.horizontalAlignment = Element.ALIGN_CENTER
                header.borderWidth = 2f
                header.phrase = Phrase(headerTitle, headFont)
                table.addCell(header)
            }
            for (employee in employees) {
                val idCell = PdfPCell(Phrase(employee.id.toString()))
                idCell.paddingLeft = 4f
                idCell.verticalAlignment = Element.ALIGN_MIDDLE
                idCell.horizontalAlignment = Element.ALIGN_CENTER
                table.addCell(idCell)
                val firstNameCell = PdfPCell(Phrase(employee.firstName))
                firstNameCell.paddingLeft = 4f
                firstNameCell.verticalAlignment = Element.ALIGN_MIDDLE
                firstNameCell.horizontalAlignment = Element.ALIGN_LEFT
                table.addCell(firstNameCell)
                val lastNameCell = PdfPCell(Phrase(employee.lastName.toString()))
                lastNameCell.verticalAlignment = Element.ALIGN_MIDDLE
                lastNameCell.horizontalAlignment = Element.ALIGN_RIGHT
                lastNameCell.paddingRight = 4f
                table.addCell(lastNameCell)
            }
            document.add(table)
            document.close()
        } catch (e: DocumentException) {
            logger.error(e.toString())
        }
        return ByteArrayInputStream(out.toByteArray())
    }
}